### PR TITLE
Hide Barchart Value Labels for Automatic Display When Overlapping

### LINKED ISSUE #42905: Bar chart overlaping numbers
Hey guys,
Screenshot_8

How to fix overlaping numbers in bar chart, data source used Infinity, Grafana version 8.1.7. Is there a way to fit numbers inside bars?

### PR DESCRIPTION
**Pull request by codeincarnate**
This PR fixes an issue where in certain instances when barchart values labels will overlap. This is especially true on data with high variability as the scaling will cause labels to easily overlap. When value labels are set to display as "Auto" this PR will hide any labels which overlap with other labels.

To accomplish this I added an additional Quadtree for labels, and added code to compare bounding box intersections to our Quadtree implementation.

**Which issue(s) this PR fixes**:

Fixes #42905

### CODE REVIEW

**Comment by leeoniya** [comment]
hey @codeincarnate a few questions:

since we're likely only checking against text labels collisions at the same x index, we can probably iterate that entire subset without constructing another quadtree? it will almost always be single digit quntity of label bounding rects.

rather than re-measuring labels each time, perhaps we can use the approximate (though not perfect) bounding boxes which we already calculate cheaply here:
 https://github.com/grafana/grafana/blob/14fe320ea8f12d9c01eeb00314d664ed433df3d2/public/app/plugins/panel/barchart/bars.ts#L343-L344

this may or may not work, since it only supplies the _available_ space, and not the final label width. internally, `calculateFontSize()` already calls `measureText()`, so feels kinda wasteful to do it twice for every label. wonder if we can copy/paste/tweak that function to also return the multiplied/computed text width. :thinking:

**Comment by codeincarnate** [comment]
Thanks for the thoughtful review!

> since we're likely only checking against text labels collisions at the same x index, we can probably iterate that entire subset without constructing another quadtree? it will almost always be single digit quntity of label bounding rects.

I'd actually taken this approach initially but it seemed a bit naive, but I'm fine with going back to it 😄 

> rather than re-measuring labels each time, perhaps we can use the approximate (though not perfect) bounding boxes which we already calculate cheaply here:
> 
> https://github.com/grafana/grafana/blob/14fe320ea8f12d9c01eeb00314d664ed433df3d2/public/app/plugins/panel/barchart/bars.ts#L343-L344


> this may not work tho, since it only supplies the _available_ space, and not the final label width. internally, `calculateFontSize()` already calls `measureText()`, so feels kinda wasteful to do it twice for every label. wonder if we can copy/paste/tweak that function to also return the multiplied/computed text width. 🤔

I'll take a look at these, I can compare the bounding boxes and see how they end up as well. Thanks!

**Comment by codeincarnate** [comment]
I tested above using the approximate values to see if that would work, however there's quite a bit more "potential" space where a label can be placed beyond it's bounding box so we end up with the following (blue is approximate bounding box vs red being bounding box from the 2d context):

Screen Shot 2022-02-03 at 12 08 07 PM

This still leaves us calling measureText more than once, however it may be necessary as the implementation in `cacluateFontSize` is called on a static font size which may then end up changing from auto-sizing. That being said I changed to using `measureText()` to take advantage of static caching in situations where that doesn't occur.

I also switched to a table oriented comparison. It's potentially n*n in complexity (with n being the number of columns) which isn't great but it seems unlikely this would get crazy

**Comment by leeoniya** [comment]
thanks for testing!

> This still leaves us calling measureText more than once, however it may be necessary as the implementation in cacluateFontSize is called on a static font size which may then end up changing from auto-sizing.

the bounding box _proportions_ won't change though, so whatever you get for 14px can be scaled up with good results, you just need to get back the raw width/height from `measureText`. the actual `calculateFontSize()` impl takes advantage of the proportions staying consistent to fit the text into the supplied limits:

https://github.com/grafana/grafana/blob/f1b2e2b7ffb7a3df136fb79c778564a24817223e/packages/grafana-ui/src/utils/measureText.ts#L40-L53

**Comment by codeincarnate** [comment]
Alright, I think I've managed to capture the all the different scenarios and necessary updates:

1. I updated `calculateFontSize` so that it returns both font size and a `TextMetrics` object. This meant some unrelated changes elsewhere but doesn't seem too serious.
2. I updated the font size calculation loop so that `measureText` will only be called once per label whether the font size is set to auto or if it's been set by the end user.
2. In the case of auto-sizing, a scale factor is calculated to adjust bounding boxes if the font-size is changed.
3. Removed `texts` array in favor of using `ValueLabelTable` exclusively.
4. Calculate bounding box adjustments based on orientation.

See below for a video of the different scenarios with bounding boxes being shown here for the sake of brevity.

https://user-images.githubusercontent.com/199847/152614251-757cdc55-6e09-4d69-b582-7ce88b8de712.mov

**Comment by codeincarnate** [comment]
Spoke with @leeoniya and I think I've identified an approach that should prevent breaking changes while generally maintaining performance. For reference, I used the following panel with ~470 data points for performance testing.

Screen Shot 2022-02-07 at 4 11 56 PM

In order to test the performance of the code I used the Chrome JS profiler and tested resizing the window with the panel displayed taking of the full size of the dashboard. When using an approximation of `measureText()` I got the following result:


### Baseline (main):

Screen Shot 2022-02-07 at 4 16 23 PM


### Approximation:

Screen Shot 2022-02-07 at 4 17 09 PM


### Direct Measurement with `measureText()` (font size set to "auto")


Screen Shot 2022-02-07 at 4 21 02 PM

Per the profiles using `measureText()` does indeed introduce significant overhead when font size is set to "auto". Thankfully the approximate approach appears to add ~0.5% to 1% runtime overhead which seems like it should be acceptable in this case. Otherwise happy for feedback on how to eek some more performance out of this.

**Comment by leeoniya** [comment]
will review this in a few hours. currently slowly working through testing a new [uPlot](https://github.com/leeoniya/uPlot) release.

also, i just discovered this profile-polluting bug in BarChart that affects resizing and mousemoves: https://github.com/grafana/grafana/pull/45027 :roll_eyes: 

**PROTIP**: `yarn build:nominify` will make an uncompressed production build, which makes for cleaner, much more realistic, and faster profiles since it strips out all devmode codepaths (i can see several listed in your profile).

**Comment by codeincarnate** [comment]
> PROTIP: yarn build:nominify will make an uncompressed production build, which makes for cleaner, much more realistic, and faster profiles since it omits all devmode codepaths (i can see several listed in your profile).

Oh man 🤦‍♂️, thanks for the catch there. I can take a look from that perspective as well. I appreciate the help and patience with this 🙏

**Comment by codeincarnate** [comment]
Retested with a production build as seeing around 10% - 11% of runtime with the approximate label bounds checking vs 6% to 7% in `main`. This seems fairly consistent with the development builds as well. I think I have a couple things that could potentially shave some of that time off. Going to see how feasible those are.

### Baseline

Screen Shot 2022-02-08 at 11 51 34 AM

### Approximate Method

Screen Shot 2022-02-08 at 11 51 41 AM

**Comment by codeincarnate** [comment]
Alright, did one more pass at this and think I've gotten it into a good position. Moved the logic for constructing label bounding boxes into the `each` callback. Because we need to have bounding boxes before drawing labels this remove a number of extra loops that processed all the rectangles being displayed.

This makes the performance measurement a bit more tricky, but I'm seeing the following:

### Baseline (main)

Screen Shot 2022-02-08 at 6 49 40 PM

Resulting runtime of draw  (6.2%) + each (1.3%) = 7.5%


### As Described

Screen Shot 2022-02-08 at 6 59 46 PM

Resulting runtime of draw (5.2%) + each (3.2%) = 8.4%


As far as I can tell the code seems to have around 1% overhead relative to the baseline implementation. That being said would be happy for further feedback in this regard or otherwise.

**Comment by leeoniya** [comment]
awesome work. perf-wise this looks good now. (now i need to fix that `.getBoundingClientRect()` situation in uPlot during panel resizes :rofl:)

think i found a bug where the font size in auto mode never recovers from the min size after data refreshes and panel resizes:

https://user-images.githubusercontent.com/43234/153122329-1907f334-f384-4b20-a740-d6f5cfeee6f3.mp4

**Comment by leeoniya** [comment]
also please take a look at the alignment of negative labels:

image

**Comment by codeincarnate** [comment]
Thanks so much Leon! I'll load up the data and work through all of these things.

**Comment by codeincarnate** [comment]
I can see the issue, currently having some trouble tracking down why exactly the end layout is different although as far as I know this code shouldn't be impacting that (most of the math is calculating bounding boxes which don't effect the end label placement). If I can't hunt it down myself will reach out for help.

**Comment by codeincarnate** [comment]
Just an update, got stuck on the negative label. Spoke with @dprokop and he said he should be able to take a look at it next Tuesday. I'm switching gears for a bit to get my head out of the code and get a little inspiration and will try to get back to it before then.

**Comment by codeincarnate** [comment]
@leeoniya I've pushed updates that fixes for negative labels and the font size becoming stuck. I've also modified the behavior to use a skip list approach. So once a label is hidden subsequent labels will ignore it for intersections.

Also did another quick run through with the JS profiler and as far as I can tell the performance is still looking good. If there's anything else please let me know!

File: public/app/plugins/panel/barchart/bars.ts [hunk 1]
```diff
@@ -355,7 +355,7 @@ export function getConfig(opts: BarsOptions, theme: GrafanaTheme2) {
           }
 
           if (baseline !== curBaseline) {
-            u.ctx.textBaseline = curBaseline = baseline;
+            u.ctx.textBaseline = baseline;
```
[the remark this answers is not in the document]
Comment (dprokop) [reply public/app/plugins/panel/barchart/bars.ts]: Thanks @leeoniya - my bad, I shhould have reverted this change, but it slipped through

**Review by leeoniya — APPROVED** [review]
awesome, LGTM :+1:

**Pull request merged**